lib/api.js

```diff
@@ -515,6 +521,62 @@ api.declare({
   res.reply(prices);
 });
 
+api.declare({
+  method: 'delete',
+  route: '/credentials',
+  name: 'claimCredentials',
+  title: 'Request worker credentials',
+  stability: API.stability.experimental,
+  input: 'claim-credentials-request.json#',
+  output: 'claim-credentials-response.json#',
+  description: [
+    'Yield credentials for the worker running in the given instance.',
+  ].join(' '),
+}, async function(req, res) {
+  const sIdentDoc = new Buffer(req.body.identityDocument, 'base64').toString();
+  const signature = new Buffer(req.body.signature, 'base64').toString();
+
+  const E = () => {
+    return res.reportError('AuthorizationFailed', 'Failure to verify authorization');
+  };
+
+  if (!await validateIdentityDocument(sIdentDoc, signature)) {
+    return E();
+  }
+
+  const identDoc = JSON.parse(sIdentDoc);
+  const instances = await this.state.listInstances({
+    id: identDoc.instanceId,
+    region: identDoc.region,
+  });
+
+  if (instances.length != 1) {
+    return E();
+  }
+
+  const instance = instances[0];
+  if (instance.state != 'running' || instance.hasClaimedCredentials) {
+    return E();
+  }
+
+  const provisionerId = this.keyPrefix.slice(0, this.keyPrefix.length - 1);
+
+  const result = res.reply(taskcluster.createTemporaryCredentials({
+    scopes: [
+      `assume:worker-type:${provisionerId}/${instance.workerType}`,
+      'assume:worker-id:*',
+    ],
```

Comment:
Yes, these scopes look good.

Reply:
I think at this point, we should also strongly consider issuing
`assume:worker-id:$region:$instanceId` in place of the star scopes here.  Since this change is going to require a change to how workers work, this might be a good time to go through with that testing.  Both because we can, and because we finally have understanding that instance ids will not be reused.

Reply:
That would be so awesome!!